cmd/bnsd: add batch support and cli tests

husio opened the pull request:
Implement batch message creation. Add `clitests` directory containing
`bnscli` test scripts.
Minor improvements.

This update fulfills requirements from #625. See
`clitests/batch_proposal.test` file for how to create batch proposal.

Quite some changes in the background:
- new `as-batch` command that can batch [**some** messages](https://github.com/iov-one/weave/compare/bnscli_improvements?expand=1#diff-1e9ac3819e63aac149ebdf57a2b9c78fR41)
- `as-proposal` command supports **some** of the batch messages
- a test suite for `bnscli` binary that provides a cookbook for how to write pipelines (see `clitests/` directory)
- only a sprinkle of documentation as it tend to get outdated

ethanfrey commented on cmd/bnscli/README.md:
```diff
@@ -0,0 +1,29 @@
+# `bnscli`
```
Nice README!

ethanfrey commented on cmd/bnscli/clitests/README.md:
```diff
@@ -0,0 +1,39 @@
+
+### Creating a golden file
+
+Do not create `xxx.test.gold` files by hand. Instead, run the test runner with
```
:+1:

husio replied:
This is an another idea from a great presentation on testing https://www.youtube.com/watch?v=yszygk1cpEc&feature=youtu.be&t=867

ethanfrey commented on cmd/bnscli/clitests/batch_proposal.test:
```diff
@@ -0,0 +1,17 @@
+bnscli send-tokens -src 'seq:foo/src/1' -dst "seq:foo/dst/1" -amount "2 IOV" -memo "sending 2 IOV" >> $msgstream
+bnscli send-tokens -src 'seq:foo/src/2' -dst "seq:foo/dst/2" -amount "9 DOGE" -memo "sending 9 DOGE" >> $msgstream
+bnscli send-tokens -src 'seq:foo/src/3' -dst "seq:foo/dst/3" -amount "7 BTC" -memo "sending 7 BTC" >> $msgstream
+bnscli release-escrow -escrow 04 >> $msgstream
```
I would place this on the top line if you want to demonstrate an expected workflow.
(Release the escrow, then send the tokens (that were just released)

ethanfrey commented on cmd/bnscli/clitests/batch_proposal.test.gold:
```diff
@@ -0,0 +1,80 @@
+						"metadata": {
+							"schema": 1
+						},
+						"escrow_id": "BA=="
```
This is an invalid ID. They are all 8 bytes bigendian integers.
I would suggest that the input takes an integer and encodes it properly.
Or else takes this hex, but enforces it to be 16 chars (if cut and paste from some output)

husio replied:
I was wondering about doing the validation of the input. I have decided that accepting most values and letting the server validate it is easier. For the user it is a similar result - the validation fails. At the same time, we do not have to hardcode a custom validation logic into the `bnscli`. `bnscli` is also more powerful as we can create invalid messages to check if `bnsd` validation is correct ;)

What do you think?

ethanfrey commented on cmd/bnscli/clitests/batch.test:
```diff
@@ -0,0 +1,12 @@
+# Combine all messages in a single file to build a batch transaction.
+bnscli send-tokens -src 'seq:foo/src/1' -dst "seq:foo/dst/1" -amount "2 IOV" -memo "sending 2 IOV" >> $msgs
+bnscli send-tokens -src 'seq:foo/src/2' -dst "seq:foo/dst/2" -amount "9 DOGE" -memo "sending 9 DOGE" >> $msgs
+bnscli send-tokens -src 'seq:foo/src/3' -dst "seq:foo/dst/3" -amount "7 BTC" -memo "sending 7 BTC" >> $msgs
```
This is very cool how you can just append them.
I guess you took advantage of the protobuf repeated trick?

husio replied:
No, protobuf cannot be used for streaming on it's own. I had to reimplement how each transaction is [written and read](https://github.com/iov-one/weave/blob/efe6ed806da868fc8fc8c46131bb0af4930c2c73/cmd/bnscli/common.go#L18-L60) to include the message size information.

https://developers.google.com/protocol-buffers/docs/techniques#streaming

ethanfrey commented on cmd/bnscli/clitests/batch.test:
```diff
@@ -0,0 +1,12 @@
+bnscli send-tokens -src 'seq:foo/src/2' -dst "seq:foo/dst/2" -amount "9 DOGE" -memo "sending 9 DOGE" >> $msgs
+bnscli send-tokens -src 'seq:foo/src/3' -dst "seq:foo/dst/3" -amount "7 BTC" -memo "sending 7 BTC" >> $msgs
+
+bnscli as-batch < $msgs | bnscli view
```
How about `rm $msgs` for cleanup at the end? to avoid filling up my /tmp directory?

ethanfrey replied:
(same for batch_proposal)

husio replied:
I can write it instead 

```sh
(
    bnscli send-tokens ...
    bnscli send-tokens ...
    ...
) | bnscli as-batch | bnscli view
```
This way I don't have to use a temporary file. I thought this is less readable so I went with temporary file.
What do you think?

ethanfrey commented on cmd/bnscli/clitests/batch_proposal.test:
```diff
@@ -0,0 +1,17 @@
+bnscli send-tokens -src 'seq:foo/src/3' -dst "seq:foo/dst/3" -amount "7 BTC" -memo "sending 7 BTC" >> $msgstream
+bnscli release-escrow -escrow 04 >> $msgstream
+
+bnscli as-batch < $msgstream \
```
Crazy powerful unix shell magic! :zap:

husio replied:
![](https://i.imgur.com/WbRowtv.jpg)

ethanfrey commented on cmd/bnscli/clitests/batch_proposal.test:
```diff
@@ -0,0 +1,17 @@
+bnscli release-escrow -escrow 04 >> $msgstream
+
+bnscli as-batch < $msgstream \
+	| bnscli as-proposal -start "2021-01-01 11:11" -electionrule 3 -title "my proposal" -description "yet another proposal" \
```
Interesting here you parse an int into the 8 byte hex, where escrow assumes those 8 bytes...

ethanfrey replied:
The 8 bytes is more general, but the integer works great with stuff defined in the genesis file (and I assume in reality most of these commands will be used with genesis-defined objects)

husio replied:
My motivation was exactly the declaration in the genesis. I agree that using hex encoded value is more powerful. The current implementation might be confusing (electionrule is an int, but escrow is a hex). Do you think I should change it to a hex value? If yes, than [validation discuss](https://github.com/iov-one/weave/pull/735#discussion_r291490219) is relevant.

ethanfrey replied:
I would actually change the escrow one, just notices you did it (in my eyes) better here.

husio replied:
I can do this, it's an easy change in both cases.

Aren't you afraid that using a number explicitly will make it harder later to provide the value? For example, when I create an escrow, I will get back a byte array. I can decode it to a number if I know it represents a number. Otherwise, I got bytes and I don't know how to convert it.

ethanfrey replied:
Yes, but if 15 is valid, it could either be one byte hex, or an int.

I would assert either:
* integer, decimal encoded
* 8 bytes, hex encoded

and then it is impossible that you intend an integer and it is interpreted as hex bytes.

I can go either way which is more correct to enter, but we should enforce the format when parsing

husio commented:
I need to fix the test runner for the `cmd/bnscli/clitests/` as travis fails due to a missing `bnscli` binary.

ethanfrey commented on cmd/bnscli/clitests/send_tokens.test:
```diff
@@ -0,0 +1,13 @@
+set -e
+
+keyfile=`tempfile`
+echo 00wZcK6QrPNAXy2Z3KyhbQx9s3n0vq/P32Z7nWnONQ0n9ftEBQnfp57Ig6BRC8mpYUw9RBiIgfDF5AKJi0vzyQ== | base64 -d > $keyfile
```
It would be great to document where you created this in order to make the test case.
If it was hard, maybe we can add a `bnscli keygen` command?
Or course, you will use this hard-coded key here, but we can point people to use `keygen` once to set up new keys for eg. genesis file

husio replied:
I will write comment while it is hardcoded and use it in a place where it is actually needed (the signature test case).

ethanfrey commented on cmd/bnscli/clitests/send_tokens.test:
```diff
@@ -0,0 +1,13 @@
+keyfile=`tempfile`
+echo 00wZcK6QrPNAXy2Z3KyhbQx9s3n0vq/P32Z7nWnONQ0n9ftEBQnfp57Ig6BRC8mpYUw9RBiIgfDF5AKJi0vzyQ== | base64 -d > $keyfile
+
+bnscli send-tokens \
```
Huh, keyfile isn't even used here. Did you forget the `sign` step?

husio replied:
`sign` test was not possible because the connection to the server is required. I have implemented this for the tests now.

ethanfrey commented on cmd/bnscli/cmd_gov.go:
```diff
@@ -67,6 +59,46 @@ transaction (ie signatures) are being dropped.
+		var messages []app.ProposalBatchMsg_Union
+		for _, m := range msgs {
+			switch m := m.(type) {
+			case *cash.SendMsg:
```
Same point I raised about batch (please as a separate PR).
Since people will not look at this logic when adding a new type, it would be good to add support for all supported types proposal batch. (Both in this case, and above in the top level case).

This logic here is pretty repetative and the type-checker will ensure you make no mistakes, so I don't think it matters if some of those cases are not tested

husio replied:
Done in https://github.com/iov-one/weave/pull/735/commits/0804afabf48a35413a0603c97d1cbbf065165380 and https://github.com/iov-one/weave/pull/735/commits/b36988931519e7ffdc52426b756b86d580383ae7

ethanfrey commented on cmd/bnscli/cmd_gov.go:
```diff
@@ -67,6 +59,46 @@ transaction (ie signatures) are being dropped.
+				Messages: messages,
+			},
+		}
+	case *app.ProposalBatchMsg:
```
I like this one... It is a noop for safety

ethanfrey replied:
But... this will never be the case for a top-level bnsd/app.Tx.
It ca be removed.

ethanfrey commented on cmd/bnscli/common.go:
```diff
@@ -9,3 +14,47 @@ func sequenceID(n uint64) []byte {
+
+// writeTx serialize the transaction using a protocol buffer. First bytes
+// written contain the information how much space the transaction takes.
+// Size information is required to be able to stream the messages:
```
Ahhh.... very nice. You don't use repeated after all. but best practices for streams in order to support batch

:tipping_hand_man:

ethanfrey commented:
Test failure on CI is weird: https://travis-ci.com/iov-one/weave/jobs/206037611#L845-L853

It works locally for me. Maybe `/bin/bash` is not present?
`go test ./cmd/bnscli/clitests/` works for me locally, as does `make test`

husio commented:
I have fixed tests in https://github.com/iov-one/weave/pull/735/commits/799488eecf894f8df1a91b7c31fd90b9e0e559b8
The issue was that test scripts in `clitests/` require `bnscli` binary to be present. 

I have updated tests to skip if the `bnscli` is not available. At the same time, I have updated `Makefile` test instruction to install `bnscli` first. This ensures the binary is available and in the latest version.

husio commented on cmd/bnscli/cmd_batch.go:
```diff
@@ -38,19 +44,116 @@ original transactions (ie signatures) are being dropped.
+		//
+		// Instead of manually managing this list, use the script from
+		// the bottom comment to generate all the cases. You are
+		// welcome.
```
:smirk:

ethanfrey replied:
Thank you for adding them all

husio commented:
@ethanfrey I have addressed all comments. All is updated except for the validation. I am not sure what is the best solution for this problem.

ethanfrey commented on Makefile:
```diff
@@ -26,6 +26,9 @@ install:
 
 test:
+	@# bnscli binary is required by some tests. In order to not skip them, ensure bnscli binary is provided and in the latest version.
+	go install -mod=readonly github.com/iov-one/weave/cmd/bnscli
```
Interesting use of full path... I usually install `./cmd/bnscli`, but I guess they both work

ethanfrey commented on cmd/bnscli/clitests/sign_send_tokens.test:
```diff
@@ -0,0 +1,28 @@
+then
+	# TM_ADDR must be either set by the host or it will be configured by
+	# the test runner to use tendermint mock server.
+	>&2 echo "TM_ADDR must be set in order for this test to run."
```
Note that when I run it, I get this line,
And also the `sign_send_tokens.test.gold` is empty (as it expects this failure).

If you want to include any `send` tests (I like it), please start up a local node for the tests to run against.
There are example scripts in iov-core and the wallet-demo and other repos.

husio replied:
The `cli_main.go` is running a fake server. I think this is enough for a fast test.

I will think about providing a real server. I know this is way more work with all the configuration, recent `bnsd` binary is required and the process cannot be controlled (sleep here and there to ensure the server is ready :see_no_evil:). 

I think that for now, using a fake tendermint server to have an instant test feedback and an easy way to write those tests is worth the tradeof. `bnscli` is meant to be used and we can always run a command manually to check it. This is a very simple thing to do, while to automate it takes a lot of effort.

ethanfrey commented on cmd/bnscli/clitests/submit_send_tokens.test.gold:
```diff
@@ -0,0 +1 @@
+exit code: 0
```
Nice, so this one did run, while the sign_send one did not?
What did you run against?

Also interesting that this is an unsigned transaction and would be rejected by CheckTx, yet `bnscli submit` considered it a success. So I think there is an issue with `bnscli submit`

husio replied:
The `bnscli` does not validate content but relies on the server to do the work.
In this case, I am running a fake server so that the result is reproducible. The server does not validate anything and does not care about the requested content. It always responds with a successful message for the requested HTTP path. That's why submitting an unsigned request succeeds.

I try to keep the tests minimal. This test is only about submitting a message. There is no great way here to ensure that submitted content is correct, only that the command works as expected (pipe a transaction to it and it will be send). The implementation details are tested in the Go unit test though.

Does this make sense?

ethanfrey commented on cmd/bnscli/cmd_batch.go:
```diff
@@ -0,0 +1,216 @@
+building. Make sure that the "protobuf" string contains the most recent
+declaration.
+
+#!/bin/bash
```
Nice inline codegen :joy:
